### PR TITLE
Creating a text near the center of a shape should put it in the center

### LINKED ISSUE #264: Creating a text near the center of a shape should put it in the center
It's very common to build a shape (rectangle, ellipse) and want to have a label inside. Right now it's hard to click precisely enough, I almost always have to recenter the text after having written it. Would be nice if we could notice that you are trying to write something at the center and automatically "snap" it to the center.

### PR DESCRIPTION
**Pull request by nanot1m**
Fixes #264 

Screen Recording 2020-01-09 at 00 26 04

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -110,6 +111,7 @@ function addTextElement(
 
 const ELEMENT_SHIFT_TRANSLATE_AMOUNT = 5;
 const ELEMENT_TRANSLATE_AMOUNT = 1;
+const TEXT_TO_CENTER_SNAP_THRASHOLD = 60;
```
Comment (vjeux) [inline src/index.tsx]: typo: thrEshold

**Comment by vjeux** [comment]
60 pixels seems a bit too much. Can you go with something smaller. I think 30px would be better

**Comment by vjeux** [comment]
Having alt active should override snap, in case the heuristic doesn't do the right thing

**Comment by vjeux** [comment]
This is awesome btw, this is going to make creating shapes a lot easier!

**Comment by nanot1m** [comment]
> Having alt active should override snap

@vjeux, sorry, I do not understand what does "Having alt active" mean 😔 Maybe some example would help

**Comment by vjeux** [comment]
If you have the alt key pressed down when you click, it should put it at the exact position you clicked and ignore the heuristic to put it in the center of a nearby shape

**Comment by nanot1m** [comment]
Thanks! Thats make sense

File: src/index.tsx [hunk 2]
```diff
@@ -970,6 +970,41 @@ class App extends React.Component<{}, AppState> {
                 elementAtPosition.y +
                 CANVAS_WINDOW_OFFSET_TOP +
                 elementAtPosition.height / 2;
+            } else if (!e.altKey) {
+              // Element must be searched from the end
+              // to pick element with the highest z-index
+              const elementClickedInside = elements
+                .slice()
+                .reverse()
```
Comment (vjeux) [inline src/index.tsx]: I would prefer not to do that, there can be a lot of elements and this is doing two expensive copy of a potentially big array. If you can write a normal loop that would be better.

**Comment by vjeux** [comment]
Feel free to merge whenever you're done

File: src/index.tsx (hunk 2 above)
Comment (dwelle) [reply src/index.tsx]: I think considering what we're doing on each render, `slice` & `reverse` (irrespective of array size) is the least of our perf concerns :), but why not.

Comment (vjeux) [reply src/index.tsx]: Perf is going to be a concern for this kind of application. We should avoid introducing obviously bad things perf wise when it doesn’t need to.

We’re not that inefficient right now, all the shape draw methods are precomputed. So we’re doing one pass to draw things and one pass for hit testing.

I don’t believe that we allocate that many things on those two passes.

We’ll want to implement some sort of frustrum culling to not draw things out of screen and make hit test faster by using bounds approximations but let’s worry about that for another day.

Comment (vjeux) [reply src/index.tsx]: For the draw commands, if we start using webgl, we can send draw commands to the gpu upfront and reuse them, so we don't have to send that entire list all the time. I've seen massive speedups when doing that.

**Pull request merged**